### PR TITLE
feat: cache port

### LINKED ISSUE #29: add redis

### PR DESCRIPTION
**Pull request by thuan2172001**
resolve #29

### CODE REVIEW

File: src/adapter/src/repositories/in_memory/cache.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        expiration: Option<Duration>,
+    ) -> Result<bool, CoreError> {
+        self.cleanup_expired_entries().await?;
+        let expiry_time = expiration.map(|exp| SystemTime::now() + exp);
+        let mut cache = self.cache.write().await;
+        cache.insert(
+            key,
+            (
+                value,
+                expiry_time.unwrap_or_else(|| SystemTime::UNIX_EPOCH),
```
Comment (zk-steve) [inline src/adapter/src/repositories/in_memory/cache.rs:85]: Wrong!?

File: src/adapter/src/repositories/in_memory/cache.rs
Comment (zk-steve) [inline src/adapter/src/repositories/in_memory/cache.rs:1]: Write test cases for this file

**Review by zk-steve — CHANGES_REQUESTED** [review]

File: src/adapter/src/repositories/in_memory/cache.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+use rust_core::ports::cache::CachePort;
+
+/// Represents an in-memory cache implementation.
+pub struct InMemoryCache {
+    cache: Arc<RwLock<HashMap<String, (String, SystemTime, Option<Duration>)>>>,
+}
+
+impl InMemoryCache {
+    /// Creates a new empty in-memory cache.
+    pub fn new() -> Self {
```
Comment (zk-steve) [inline src/adapter/src/repositories/in_memory/cache.rs:20]: impl Default instead of new

File: src/adapter/src/repositories/in_memory/cache.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the key is found and removed, otherwise returns a `CoreError`.
+    async fn remove(&mut self, key: &str) -> Result<bool, CoreError> {
+        let mut cache = self.cache.write().await;
+        if cache.remove(key).is_some() {
+            Ok(true)
+        } else {
+            Err(CoreError::NotFound)
```
Comment (zk-steve) [inline src/adapter/src/repositories/in_memory/cache.rs:106]: Not match the behavior of Redis `Ok(false)`.

File: src/adapter/src/repositories/redis/cache.rs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    /// Retrieves a value from the Redis cache.
+    ///
+    /// # Arguments
+    ///
+    /// * `key`: The key of the value to retrieve.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result` containing the value associated with the key if found, or `None`.
+    async fn get(&self, key: &str) -> Result<Option<String>, CoreError> {
```
Comment (zk-steve) [inline src/adapter/src/repositories/redis/cache.rs:43]: Treat empty as NotFound errror (for both version)

**Comment by zk-steve** [comment]
- [ ] Update readme

File: src/adapter/src/repositories/in_memory/cache.rs (hunk 3 above)
Comment (thuan2172001) [reply src/adapter/src/repositories/in_memory/cache.rs:111]: I want the `NotFound` error has same meaning to `false` response.

File: src/adapter/src/repositories/in_memory/cache.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    /// Removes a key-value pair from the cache based on the provided key.
+    ///
+    /// # Arguments
+    ///
+    /// * `key`: A string representing the key to remove from the cache.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the key is found and removed, otherwise returns a `CoreError`.
```
[the remark this answers is not in the document]
Comment (thuan2172001) [reply src/adapter/src/repositories/in_memory/cache.rs:105]: If there are no errors, it just means `Ok(())` in rust and  `OK` in Redis.

File: src/adapter/src/repositories/redis/cache.rs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    /// Removes a key-value pair from the Redis cache.
+    ///
+    /// # Arguments
+    ///
+    /// * `key`: The key to remove.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result` where `Ok(())` indicates success (key was removed) and `Err(CoreError)` indicates failure.
+    async fn del(&mut self, key: &str) -> Result<(), CoreError> {
```
Comment (zk-steve) [inline src/adapter/src/repositories/redis/cache.rs:94]: async fn del(&mut self, key: &str) -> Result<(), CoreError> {
        self.client
            .send(resp_array!["DEL", key])
            .await
            .map_err(|e| CoreError::InternalError(e.into()))
            .and_then(|RespValue::Integer(num_deleted)| {
                if num_deleted >= 1 {
                    // At least one key was deleted
                    Ok(())
                } else {
                    // Key does not exist
                    Err(CoreError::NotFound)
                }
            })
    }

Comment (zk-steve) [reply src/adapter/src/repositories/redis/cache.rs:94]: if num_deleted > 0 {
                    // At least one key was deleted
                    Ok(())
                } else {
                    // Key does not exist
                    Err(CoreError::NotFound)
                }
```suggestion
    async fn del(&mut self, key: &str) -> Result<(), CoreError> {
```

File: src/gpt_answer_server/src/options.rs [hunk 7]
```diff
@@ -15,7 +15,16 @@ pub struct Options {
...
+    pub redis: RedisConfig,
     /// Configuration for logging, including log level.
     #[serde(default = "default_log")]
     pub log: Log,
 }
+
+/// Represents redis configuration.
+#[derive(Debug, Deserialize, Clone)]
+pub struct RedisConfig {
+    pub host: String,
```
Comment (zk-steve) [inline src/gpt_answer_server/src/options.rs:28]: Comments

File: src/gpt_answer_server/src/controllers/gpt_answer.rs [hunk 8]
```diff
@@ -44,9 +46,10 @@ impl GptAnswerService for GptAnswerServiceImpl {
         // Extract the payload containing the question from the request
         let payload = request.into_inner();
 
-        // TODO: Implement your logic to generate an answer based on the question.
-        // Placeholder logic: Generate an answer string
-        let answer = format!("Answer to: {}, {}", payload.question, self.dummy_prop);
+        let answer =
+            self.cache.get(&payload.question).await.map_err(|err| {
+                Status::internal(format!("Failed to get answer from cache: {}", err))
```
Comment (zk-steve) [inline src/gpt_answer_server/src/controllers/gpt_answer.rs:51]:
```suggestion
                Status::internal(format!("failed to get answer from cache: {}", err))
```

File: src/core/src/ports/cache.rs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+pub trait CachePort {
+    /// Retrieves a value from the cache based on the given key.
+    ///
+    /// # Arguments
+    ///
+    /// * `key`: The key to look up in the cache.
+    ///
+    /// # Returns
+    ///
+    /// Returns an optional value associated with the key, or `None` if the key is not present in the cache.
```
Comment (zk-steve) [inline src/core/src/ports/cache.rs:18]: Comments

File: src/adapter/src/repositories/repository_test.rs [hunk 10]
```diff
@@ -126,3 +126,109 @@ mod tests {
...
+                "key1".to_string(),
+                "value1".to_string(),
+                Some(Duration::from_secs(2)),
+            )
+            .await
+            .unwrap();
+        assert_eq!(cache.get("key1").await.unwrap(), "value1".to_string());
+
+        // Test expiration
+        sleep(Duration::from_secs(3)).await; // Wait for 3 seconds to let the entry expire
```
Comment (zk-steve) [inline src/adapter/src/repositories/repository_test.rs:157]: Need further investigate

Comment (thuan2172001) [reply src/adapter/src/repositories/repository_test.rs]: running test in parallel

File: src/gpt_answer_server/src/controllers/gpt_answer.rs
Comment (zk-steve) [inline src/gpt_answer_server/src/controllers/gpt_answer.rs:1]: Fix this file

File: Makefile [hunk 11]
```diff
@@ -28,12 +28,12 @@ build:
 	PKG_NAME=rust-api-server
 	BUILD_VERSION=$(shell git describe --long)
 	BUILD_RELEASE=$(shell git describe --tags --abbrev=0)
-	BUILDKIT_PROGRESS=plain
-	DOCKER_BUILDKIT=1
-	docker build --ssh default -t $(PKG_NAME):$(BUILD_VERSION) --target=prod .
+	BUILDKIT_PROGRESS=plain	docker build --ssh default -t $(PKG_NAME):$(BUILD_VERSION) --target=public_server_prod .
```
Comment (zk-steve) [inline Makefile:31]:
```suggestion
	BUILDKIT_PROGRESS=plain	docker build -t $(PKG_NAME):$(BUILD_VERSION) --target=public_server_prod .
```
remove it in all places

File: src/adapter/src/repositories/redis/cache.rs [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+                } else {
+                    // Key does not exist
+                    Err(CoreError::NotFound)
+                }
+            }
+            Err(e) => {
+                // Error in sending command
+                Err(CoreError::InternalError(e.into()))
+            }
+            _ => Err(CoreError::NotFound),
```
Comment (zk-steve) [inline src/adapter/src/repositories/redis/cache.rs:109]: NotFound could be misleading

**Review by zk-steve — APPROVED** [review]

File: src/adapter/src/repositories/redis/cache.rs (hunk 12 above)
Comment (thuan2172001) [reply src/adapter/src/repositories/redis/cache.rs]: `
self.client
            .send(resp_array!["DEL", key])
            .await
            .map_err(|e| CoreError::InternalError(e.into()))
            .and_then(|resp| match resp {
                RespValue::Integer(num_deleted) => {
                    if num_deleted > 0 {
                        Ok(())
                    } else {
                        Err(CoreError::NotFound)
                    }
                }
                response => Err(CoreError::UnexpectedResponse(String::from(format!(
                    "Expect `RespValue::Integer` but found {:?}",
                    response
                )))),
            })
      `

File: README.md [hunk 13]
```diff
@@ -98,23 +98,26 @@ Hierarchical child config via env, separated by using `__`. Specify list values
...
+| REDIS\_\_PORT                                                          | 6379          |           |
 
 Make sure to set these environment variables according to your needs before running the server.
 
 ## GitHub Flow CI Configuration
 
 1. **Set Docker Hub Secrets:**
-    - Go to repository Settings > Secrets.
-    - Add `DOCKER_USERNAME` and `DOCKERHUB_TOKEN`.
+
```
Comment (zk-steve) [inline README.md:115]: Remove this line

Comment (thuan2172001) [reply README.md:115]: Formatted by **prettier**. It does not affect the file in view mode.

**Review by zk-steve — APPROVED** [review]

**Pull request merged**